File: 26_Dynamic Programming-II/M-TUF-Job sequencing Problem.java
// Job sequencing Problem --> Maximum profit in Job scheduling
// https://takeuforward.org/plus/dsa/problems/job-sequencing-problem

/**
 * 
 *   Given an 2D array Jobs of size Nx3, where Jobs[i][0] represents JobID , Jobs[i][1] represents Deadline , Jobs[i][2] represents Profit associated with that job. Each Job takes 1 unit of time to complete and only one job can be scheduled at a time.
 *
 *   The profit associated with a job is earned only if it is completed by its deadline. Find the number of jobs and maximum profit.
 *   
 *   Examples:
 *   Input : Jobs = [ [1, 4, 20] , [2, 1, 10] , [3, 1, 40] , [4, 1, 30] ]
 *   Output : 2 60
 *   Explanation : Job with JobID 3 can be performed at time t=1 giving a profit of 40.
 *   
 *   Job with JobID 1 can be performed at time t=2 giving a profit of 20.
 *   No more jobs can be scheduled, So total Profit = 40 + 20 => 60.
 *   Total number of jobs completed are two, JobID 1, JobID 3.
 *   So answer is 2 60. ̰
 */

import java.util.*;

class Solution {
    // Memoization table: dp[t] = maximum profit that can be achieved by time t
    private int[] dp;
    
    // Function to find the maximum profit and the number of jobs done using DP
    public int[] JobScheduling(int[][] jobs) {
        int n = jobs.length;
        if (n == 0) return new int[]{0, 0};
        
        // Sort jobs by deadline to process them in order
        Arrays.sort(jobs, (a, b) -> a[1] - b[1]);
        
        // Find maximum deadline
        int maxDeadline = 0;
        for (int[] job : jobs) {
            maxDeadline = Math.max(maxDeadline, job[1]);
        }
        
        // dp[t] represents the maximum profit that can be achieved by time t
        dp = new int[maxDeadline + 1];
        
        // Process each job
        for (int[] job : jobs) {
            int deadline = job[1];
            int profit = job[2];
            
            // For each time slot from deadline down to 1
            for (int t = deadline; t >= 1; t--) {
                // If we can include this job at time t
                if (dp[t] < dp[t - 1] + profit) {
                    dp[t] = dp[t - 1] + profit;
                }
            }
        }
        
        // Find the maximum profit and count the number of jobs
        int jobCount = 0;
        int currentProfit = 0;
        
        for (int t = 1; t <= maxDeadline; t++) {
            if (dp[t] > currentProfit) {
                currentProfit = dp[t];
                jobCount++;
            }
        }
        
        return new int[]{jobCount, currentProfit};
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Test Case 1: Example from problem statement
        int[][] jobs1 = {{1, 4, 20}, {2, 1, 10}, {3, 1, 40}, {4, 1, 30}};
        int[] result1 = sol.JobScheduling(jobs1);
        System.out.println("Test Case 1: " + Arrays.toString(result1));
        // Expected: [2, 60] (2 jobs with total profit 60)
        
        // Test Case 2: All jobs have same deadline
        int[][] jobs2 = {{1, 2, 100}, {2, 2, 80}, {3, 2, 60}, {4, 2, 40}, {5, 2, 20}};
        int[] result2 = sol.JobScheduling(jobs2);
        System.out.println("Test Case 2: " + Arrays.toString(result2));
        // Expected: [2, 180] (2 jobs with total profit 180)
        
        // Test Case 3: All jobs can be scheduled
        int[][] jobs3 = {{1, 5, 50}, {2, 1, 10}, {3, 2, 20}, {4, 3, 30}, {5, 4, 40}};
        int[] result3 = sol.JobScheduling(jobs3);
        System.out.println("Test Case 3: " + Arrays.toString(result3));
        // Expected: [5, 150] (all 5 jobs scheduled)
        
        // Test Case 4: Empty input
        int[][] jobs4 = {};
        int[] result4 = sol.JobScheduling(jobs4);
        System.out.println("Test Case 4: " + Arrays.toString(result4));
        // Expected: [0, 0] (no jobs)
        
        // Test Case 5: Single job
        int[][] jobs5 = {{1, 1, 100}};
        int[] result5 = sol.JobScheduling(jobs5);
        System.out.println("Test Case 5: " + Arrays.toString(result5));
        // Expected: [1, 100] (single job with profit 100)
    }
}

File: JavaScript/26_Dynamic Programming-II/M-TUF-Job sequencing Problem.js
/**
 * TUF: Job Sequencing Problem
 * https://takeuforward.org/plus/dsa/problems/job-sequencing
 *
 * Given jobs with deadlines and profits, find maximum profit
 * if only one job can be scheduled at a time.
 *
 * Time Complexity: O(N log N + N * maxDeadline) – Sort + Greedy assignment
 * Space Complexity: O(maxDeadline) – Slot array
 */

/**
 * Job Sequencing - Greedy approach
 * @param {Object[]} jobs - Array of {id, deadline, profit}
 * @returns {Object} - {count, maxProfit}
 */
function jobSequencing(jobs) {
    // Sort by profit in descending order
    jobs.sort((a, b) => b.profit - a.profit);
    
    // Find maximum deadline
    let maxDeadline = 0;
    for (const job of jobs) {
        maxDeadline = Math.max(maxDeadline, job.deadline);
    }
    
    // Slots to track scheduled jobs (1-indexed for convenience)
    const slots = new Array(maxDeadline + 1).fill(false);
    
    let count = 0;
    let maxProfit = 0;
    
    for (const job of jobs) {
        // Find a free slot for this job (as late as possible)
        for (let t = job.deadline; t > 0; t--) {
            if (!slots[t]) {
                slots[t] = true;
                count++;
                maxProfit += job.profit;
                break;
            }
        }
    }
    
    return { count, maxProfit };
}

/**
 * Disjoint Set Union for optimized job sequencing
 */
function createDSU(n) {
    const parent = Array.from({ length: n + 1 }, (_, i) => i);

    function find(x) {
        if (parent[x] !== x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    function union(x, y) {
        parent[find(x)] = find(y);
    }

    return { find, union };
}

/**
 * Job Sequencing with Union-Find for O(N log N)
 */
function jobSequencingOptimized(jobs) {
    jobs.sort((a, b) => b.profit - a.profit);
    
    let maxDeadline = 0;
    for (const job of jobs) {
        maxDeadline = Math.max(maxDeadline, job.deadline);
    }
    
    const dsu = createDSU(maxDeadline);
    
    let count = 0;
    let maxProfit = 0;
    
    for (const job of jobs) {
        // Find the latest available slot using DSU
        const availableSlot = dsu.find(job.deadline);
        
        if (availableSlot > 0) {
            // Assign job to this slot
            dsu.union(availableSlot, availableSlot - 1);
            count++;
            maxProfit += job.profit;
        }
    }
    
    return { count, maxProfit };
}

// Test cases
const jobs1 = [
    { id: 1, deadline: 4, profit: 20 },
    { id: 2, deadline: 1, profit: 10 },
    { id: 3, deadline: 1, profit: 40 },
    { id: 4, deadline: 1, profit: 30 }
];

console.log("Job Sequencing Result:", jobSequencing([...jobs1]));
// Expected: { count: 2, maxProfit: 60 }

const jobs2 = [
    { id: 1, deadline: 2, profit: 100 },
    { id: 2, deadline: 1, profit: 19 },
    { id: 3, deadline: 2, profit: 27 },
    { id: 4, deadline: 1, profit: 25 },
    { id: 5, deadline: 3, profit: 15 }
];

console.log("Job Sequencing Optimized:", jobSequencingOptimized([...jobs2]));
// Expected: { count: 3, maxProfit: 142 }

const jobs3 = [
    { id: 1, deadline: 3, profit: 35 },
    { id: 2, deadline: 4, profit: 30 },
    { id: 3, deadline: 4, profit: 25 },
    { id: 4, deadline: 2, profit: 20 },
    { id: 5, deadline: 3, profit: 15 },
    { id: 6, deadline: 1, profit: 12 },
    { id: 7, deadline: 2, profit: 5 }
];

console.log("Job Sequencing (7 jobs):", jobSequencing([...jobs3]));
